provider/ovh.go

```diff
@@ -0,0 +1,354 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/golang/sync/errgroup"
+	"github.com/ovh/go-ovh/ovh"
+	log "github.com/sirupsen/logrus"
+
+	"sigs.k8s.io/external-dns/endpoint"
+	"sigs.k8s.io/external-dns/plan"
+)
+
+const (
+	ovhDefaultTTL = 0
+)
+
+const (
+	ovhCreate = iota
+	ovhDelete
+)
+
+var (
+	// ErrRecordToMutateNotFound when ApplyChange has to update/delete and didn't found the record in the existing zone (Change with no record ID)
+	ErrRecordToMutateNotFound = errors.New("record to mutate not found in current zone")
+	// ErrNoDryRun No dry run support for the moment
+	ErrNoDryRun = errors.New("dry run not supported")
+)
+
+// OVHProvider is an implementation of Provider for OVH DNS.
+type OVHProvider struct {
+	client ovhClient
+
+	domainFilter DomainFilter
+	DryRun       bool
+}
+
+type ovhClient interface {
+	Post(string, interface{}, interface{}) error
```

Comment:
would be better to have a specific type like a struct which you already have defined

Reply:
This interface need to be compatible with the go-ovh client.
(https://github.com/ovh/go-ovh/blob/master/ovh/ovh.go#L150)
I'm using it to be able to mock the client in tests.